File: source/read_input.f90
SUBROUTINE read_input

USE matrices_to_be_shared
implicit none
include 'mpif.h'

CHARACTER (LEn=80) :: CHARSTRING,cs,cs2
integer::i,j,k,jj,qq
double precision::disi


if(rank==0)then

   open(unit=12, file="oldquit.dat", status="old", action="readwrite",   &
        position="rewind", iostat=jj)
   if(jj .ne. 0)then
      call biout('Could not open restart file!')
   end if
   read(12,*)M
   read(12,*)ne
   read(12,*)nst
   read(12,*)efs
   
end if

call MPI_BCAST(m,1,MPI_INTEGER,0,MPI_COMM_WORLD,ierr)
call MPI_BCAST(ne,1,MPI_INTEGER,0,MPI_COMM_WORLD,ierr)
call MPI_BCAST(nst,1,MPI_INTEGER,0,MPI_COMM_WORLD,ierr)
call MPI_BCAST(efs,1,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD,ierr)


skp=3*ne
svh=ne*(ne+1)/2
siv=m*((ne*(ne+1)/2)+3*ne)




!***********************************************************************************************

ALLOCATE(ncv(ne+1),scvv(nst),mass(ne,ne),ki(M,M),po(M,M),   &
         hh(M,M),ss(M,M),   &
         C(M),S(M,M),H(M,M),grad(m*(svh+skp)),   &
         dss(m*(svh+skp),m),dhh(m*(svh+skp),m),  &
  			CO(M),iv(siv),sml(ne,ne,nst),sms(skp,skp,nst),	&
  			DS(M*(svh+skp),M),DH(M*(svh+skp),M),Kinetic(M,M),Potential(M,M))



if(rank==0)then
   read(12,*)( ( mass(i,j), j=1,ne), i=1,ne)
end if

call MPI_BCAST(mass,ne*ne,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD,ierr)


if(rank==0)then
   do i=1,ne+1
      read(12,*)ncv(i)
   end do
end if

call MPI_BCAST(ncv,ne+1,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD,ierr)




if(rank==0)then
do i=1,siv
   read(12,*)iv(i)
end do
end if

call MPI_BCAST(iv,siv,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD,ierr)


if(rank==0)then
   do i=1,m
      read(12,*)C(i)
   end do
end if

call MPI_BCAST(c,m,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD,ierr)

if(rank==0)then
do i=1,nst
   read(12,*)scvv(i)
end do
end if

call MPI_BCAST(scvv,nst,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD,ierr)

!write(*,*)scvv
scvv=scvv/nst
!write(*,*)scvv
!stop

if(rank==0)then
do k=1,nst
   !write(*,*)k
   read(12,*)( ( sml(i,j,k), j=1,ne), i=1,ne)
end do
end if

call MPI_BCAST(sml,ne*ne*nst,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD,ierr)


if(rank==0)then
do k=1,nst
   read(12,*)( ( sms(i,j,k), j=1,skp), i=1,skp)
end do
end if

call MPI_BCAST(sms,skp*skp*nst,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD,ierr)


!write(*,*)spm
if(rank==0)then
close(12)
end if



END SUBROUTINE read_input

File: python/read_input.py
"""
Read the input file (oldquit.dat) and populate the shared global state.

Analogous to read_input.f90.  MPI broadcast calls have been removed; all
data is read on a single process.
"""

from __future__ import annotations

import numpy as np

import shared as sh
from utils import biout


def read_input(filename: str = "oldquit.dat") -> None:
    """
    Read *filename* and populate every field in *shared*.

    File layout (one value or row of values per line)::

        M
        ne
        nst
        efs
        mass(1,1) mass(1,2) ... mass(1,ne)
        mass(2,1) ...
        ...
        mass(ne,1) ...
        ncv(1)
        ncv(2)
        ...
        ncv(ne+1)
        iv(1)
        iv(2)
        ...
        iv(siv)
        C(1)
        ...
        C(M)
        scvv(1)
        ...
        scvv(nst)
        sml(1,1,1) sml(1,2,1) ... sml(1,ne,1)   ! row 1 of sml(:,:,1)
        sml(2,1,1) ...
        ...                                         ! all ne rows for k=1
        sml(1,1,2) ...                              ! k=2
        ...                                         ! repeat for all nst
        sms(1,1,1) sms(1,2,1) ... sms(1,skp,1)
        sms(2,1,1) ...
        ...                                         ! all skp rows for k=1
        ...                                         ! repeat for all nst
    """
    try:
        with open(filename, "r", encoding="utf-8") as fh:
            tokens = fh.read().split()
    except OSError:
        biout(f"Could not open input file: {filename}")
        raise

    pos = 0

    def _next_int() -> int:
        nonlocal pos
        val = int(tokens[pos])
        pos += 1
        return val

    def _next_float() -> float:
        nonlocal pos
        val = float(tokens[pos].lower().replace('d', 'e'))
        pos += 1
        return val

    # ------------------------------------------------------------------
    # Scalar parameters
    # ------------------------------------------------------------------
    sh.M = _next_int()
    sh.ne = _next_int()
    sh.nst = _next_int()
    sh.efs = _next_float()

    sh.skp = 3 * sh.ne
    sh.svh = sh.ne * (sh.ne + 1) // 2
    sh.siv = sh.M * (sh.svh + sh.skp)

    # ------------------------------------------------------------------
    # Allocate all arrays
    # ------------------------------------------------------------------
    sh.mass = np.zeros((sh.ne, sh.ne))
    sh.ki = np.zeros((sh.M, sh.M))
    sh.po = np.zeros((sh.M, sh.M))
    sh.hh = np.zeros((sh.M, sh.M))
    sh.ss = np.zeros((sh.M, sh.M))
    sh.C = np.zeros(sh.M)
    sh.S = np.zeros((sh.M, sh.M))
    sh.H = np.zeros((sh.M, sh.M))
    sh.Kinetic = np.zeros((sh.M, sh.M))
    sh.Potential = np.zeros((sh.M, sh.M))
    sh.grad = np.zeros(sh.siv + sh.M)
    sh.DS = np.zeros((sh.M * (sh.svh + sh.skp), sh.M))
    sh.DH = np.zeros((sh.M * (sh.svh + sh.skp), sh.M))
    sh.DSS = np.zeros((sh.M * (sh.svh + sh.skp), sh.M))
    sh.DHH = np.zeros((sh.M * (sh.svh + sh.skp), sh.M))
    sh.ncv = np.zeros(sh.ne + 1)
    sh.scvv = np.zeros(sh.nst)
    sh.iv = np.zeros(sh.siv)
    sh.sml = np.zeros((sh.ne, sh.ne, sh.nst))
    sh.sms = np.zeros((sh.skp, sh.skp, sh.nst))

    # ------------------------------------------------------------------
    # Mass matrix  (ne × ne, stored row-by-row in input file)
    # ------------------------------------------------------------------
    for i in range(sh.ne):
        for j in range(sh.ne):
            sh.mass[i, j] = _next_float()

    # ------------------------------------------------------------------
    # Charge values ncv(1..ne+1)
    # ------------------------------------------------------------------
    for i in range(sh.ne + 1):
        sh.ncv[i] = _next_float()

    # ------------------------------------------------------------------
    # Nonlinear parameter vector iv(1..siv)
    # ------------------------------------------------------------------
    for i in range(sh.siv):
        sh.iv[i] = _next_float()

    # ------------------------------------------------------------------
    # Linear coefficients C(1..M)
    # ------------------------------------------------------------------
    for i in range(sh.M):
        sh.C[i] = _next_float()

    # ------------------------------------------------------------------
    # Symmetry weights scvv(1..nst)  — normalised by nst
    # ------------------------------------------------------------------
    for i in range(sh.nst):
        sh.scvv[i] = _next_float()
    sh.scvv /= sh.nst

    # ------------------------------------------------------------------
    # Symmetry matrix sml(ne, ne, nst)  — stored row-by-row per k
    # ------------------------------------------------------------------
    for k in range(sh.nst):
        for i in range(sh.ne):
            for j in range(sh.ne):
                sh.sml[i, j, k] = _next_float()

    # ------------------------------------------------------------------
    # Symmetry matrix for shifts sms(skp, skp, nst)
    # ------------------------------------------------------------------
    for k in range(sh.nst):
        for i in range(sh.skp):
            for j in range(sh.skp):
                sh.sms[i, j, k] = _next_float()
